Incorrect object safety hints with GAT lifetimes

rust-analyzer incorrectly hints that a trait with an associated type that is generic over a lifetime is object safe, which conflicts with rustc saying the trait is not objects safe.

**rust-analyzer version**: 0.4.2111-standalone (94b526fc8 2024-09-15)

**rustc version**: rustc 1.81.0 (eeb90cda1 2024-09-04)

**editor or extension**: VSCode rust-analyzer v0.4.2114 (pre-release)

**code snippet to reproduce**: 
```rust
trait Foo {
    type Bar<'a>;
}

fn main() {
    let baz: &dyn Foo = todo!();
}
```
When hovering `trait Foo` (#17779), rust-analyzer claims `// Object Safety: Yes`; however, compiling this code yields this error which explicitly says `trait Foo` is not object safe:
```
https://doc.rust-lang.org/reference/items/traits.html#object-safety>
 --> src/main.rs:2:10
  |
1 | trait Foo {
  |       --- this trait cannot be made into an object...
2 |     type Bar<'a>;
  |          ^^^ ...because it contains the generic associated type `Bar`
  = help: consider moving `Bar` to another trait

error[E0038]: the trait `Foo` cannot be made into an object
 --> src/main.rs:6:14
  |
6 |     let baz: &dyn Foo = todo!();
  |              ^^^^^^^^ `Foo` cannot be made into an object
  |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
 --> src/main.rs:2:10
  |
1 | trait Foo {
  |       --- this trait cannot be made into an object...
2 |     type Bar<'a>;
  |          ^^^ ...because it contains the generic associated type `Bar`
  = help: consider moving `Bar` to another trait

For more information about this error, try `rustc --explain E0038`.
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"ChayimFriedman2"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/crates/hir-ty/src/object_safety.rs
+++ b/crates/hir-ty/src/object_safety.rs
@@ -349,7 +349,7 @@ where
                 ControlFlow::Continue(())
             } else {
                 let generic_params = db.generic_params(item.into());
-                if generic_params.len_type_or_consts() > 0 {
+                if !generic_params.is_empty() {
                     cb(ObjectSafetyViolation::GAT(it))
                 } else {
                     ControlFlow::Continue(())

--- a/crates/hir-ty/src/object_safety/tests.rs
+++ b/crates/hir-ty/src/object_safety/tests.rs
@@ -378,3 +378,16 @@ pub trait Error: core::fmt::Debug + core::fmt::Display {
         [("Error", vec![])],
     );
 }
+
+#[test]
+fn lifetime_gat_is_object_unsafe() {
+    check_object_safety(
+        r#"
+//- minicore: dispatch_from_dyn
+trait Foo {
+    type Bar<'a>;
+}
+"#,
+        [("Foo", vec![ObjectSafetyViolationKind::GAT])],
+    );
+}

